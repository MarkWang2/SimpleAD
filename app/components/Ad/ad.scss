.ad-placeholder {
  display: flex;
  align-items: flex-start;
  justify-content: center;
  width: 100%;
  overflow: hidden;
  line-height: 0;
}

