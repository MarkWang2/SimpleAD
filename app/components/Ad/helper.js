// import slotsConfig from './slotsConfig'

export const getSlotConfigById = (id) => {
  const { slots, devices } = window.slotsConfig

  return { devices, ...slots.find(slot => slot?.name === id) }
}