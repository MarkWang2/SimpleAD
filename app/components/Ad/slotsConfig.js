const slotsConfig = {
  "slots": [
    {
      "name": "lead2",
      "adUnit": "lead2",
      "slotTargeting": [
        {
          "name": "sdf",
          "value": "432"
        },
        {
          "name": "df",
          "value": "dsfsdf"
        }
      ],
      "sizeMapping": {
        "sm": [
          "300x250"
        ],
        "md": [
          "320x250"
        ],
        "lg": [
          "300x600"
        ]
      }
    },
    {
      "name": "cxzxc",
      "adUnit": "zxc",
      "slotTargeting": [],
      "sizeMapping": {
        "sm": [
          "df",
          "sdf"
        ],
        "md": [
          "sdf"
        ],
        "lg": [
          "sdf"
        ]
      }
    }
  ],
  "devices": {
    "lg": [
      1024,
      0
    ],
    "md": [
      767,
      0
    ],
    "sm": [
      0,
      0
    ]
  }
}
export default slotsConfig