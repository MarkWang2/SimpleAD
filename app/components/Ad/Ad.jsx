'use client'

import React, { useEffect } from 'react'
import classNames from 'classnames'
import { getSlotConfigById } from './helper'

import './ad.scss'
import AdModel from '@/app/models/adModel'

const Ad = (props) => {
  const {
    id, position, isOOP, className, adUnit, pageType,
  } = props
  const buildSlot = () => {
    const configSlot = getSlotConfigById(id, pageType)
    return configSlot || {}
  }

  const adSlot = buildSlot()
  const prebidUnits = []

  const getSizeMapping = () => {
    const devices = AdModel.getDevices().map((device) => device.split('x').map((str) => Number(str)))
    const adsSizeMapping = googletag.sizeMapping()
    devices.forEach((device, index) => {
      adsSizeMapping.addSize(device, adSlot.sizes[index])
    })
    return adsSizeMapping.build()
  }

  const defineSlot = () => {
    const defaultRequestSize = '330x250'
    const responsiveAdSlot = googletag.defineSlot(adUnit, defaultRequestSize, id).
      addService(googletag.pubads()).
      setTargeting('position', position).
      setCollapseEmptyDiv(true)
    responsiveAdSlot.defineSizeMapping(getSizeMapping(adSlot))

    return responsiveAdSlot
  }

  const defineOOP = () => googletag.defineOutOfPageSlot(adUnit, position).
    setTargeting('position', position).
    addService(googletag.pubads())

  const refreshWithBid = (slot) => {
    pbjs.que.push(() => {
      pbjs.addAdUnits(prebidUnits)
      pbjs.requestBids({
        timeout: 2000,
        bidsBackHandler () {
          googletag.cmd.push(() => {
            pbjs.que.push(() => {
              pbjs.setTargetingForGPTAsync()
              googletag.pubads().refresh([slot], { changeCorrelator: false })
            })
          })
        },
      })
    })
  }

  const pushAdSlot = () => {
    if (typeof window !== undefined) {
      const { googletag } = window
      googletag.cmd = googletag.cmd || []

      googletag.cmd.push(() => {
        debugger
        const slot = isOOP ? defineOOP() : defineSlot()
        googletag.pubads().disableInitialLoad()
        googletag.pubads().enableSingleRequest()
        googletag.enableServices()
        googletag.pubads().enableLazyLoad()
        // setPrebidConfig();
        googletag.display(slot)
        if (adSlot?.isBid) {
          refreshWithBid(slot)
        } else googletag.pubads().refresh([slot], { changeCorrelator: false })
      })
    }
  }

  useEffect(pushAdSlot, [])

  return (
    <div
      data-testid={id}
      id={id}
      key={id}
      className={classNames('ad-placeholder', className)}
      data-position={position}>
    </div>
  )
}

export default Ad