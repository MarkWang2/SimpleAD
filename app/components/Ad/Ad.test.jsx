import React from 'react';
import { html } from 'js-beautify';
import { renderWithProviders } from 'common/utils/ut';
import Ad from './Ad';

jest.mock('common/hooks/useActivePage', () => jest.fn());

class Im {
  static addService = () => {
    return this;
  };

  static setTargeting = () => {
    return this;
  };

  static setCollapseEmptyDiv = () => {
    return this;
  };

  static addSize = (device, size) => {
    googletag.slot.size = { device, size };
    return this;
  };

  static build = () => {
    return this;
  };

  static defineSizeMapping = () => {
    return this;
  };

  static disableInitialLoad = () => {
    return this;
  };

  static enableSingleRequest = () => {
    return this;
  };

  static enableLazyLoad = () => {
    return this;
  };

  static refresh = () => {
    return this;
  };
}
googletag.cmd.push = (method) => {
  method();
};
googletag.defineSlot = (adUnit, defaultRequestSize, position) => {
  googletag.slot = {
    adUnit, defaultRequestSize, position, size: '',
  };
  return Im;
};
googletag.defineOutOfPageSlot = (adUnit, defaultRequestSize, position) => {
  googletag.slot = {
    adUnit, defaultRequestSize, position, size: '',
  };
  return Im;
};

googletag.enableServices = () => {
  return Im;
};

googletag.pubads = () => {
  return Im;
};

googletag.sizeMapping = () => {
  return Im;
};

googletag.display = () => {
  return Im;
};

googletag.slot = {};

global.pbjs = {
  que: [],
};

pbjs.que.push = (method) => {
  method();
};
pbjs.aliasBidder = () => {

};

describe('<Ad />', () => {
  it('should render the ad with correct attr when set id and position', () => {
    const { container } = renderWithProviders(<Ad id='lead2' position="lead2" />, {
      preloadedState: {
        ad: { pageUnit: 'test' },
      },
    });
    expect(googletag.slot).toEqual({
      adUnit: '/21719121593/ACT/test',
      defaultRequestSize: '330x250',
      position: 'lead2',
      size: {
        device: [ 1024, 0 ],
        size: [
          [ 728, 90 ],
          [ 970, 90 ],
        ],
      },
    });
    expect(html(container.outerHTML)).toMatchSnapshot();
  });

  it('should render the ad with correct attr when is oop ad', () => {
    const { container } = renderWithProviders(<Ad isOOP id='OOP1' position="OOP1" />, {
      preloadedState: {
        ad: { pageUnit: 'test' },
      },
    });
    expect(googletag.slot).toEqual({
      adUnit: '/21719121593/ACT/test',
      defaultRequestSize: 'OOP1',
      size: '',
    });
    expect(html(container.outerHTML)).toMatchSnapshot();
  });
});