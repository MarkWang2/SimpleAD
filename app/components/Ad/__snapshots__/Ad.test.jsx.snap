// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Ad /> should render the ad with correct attr when is oop ad 1`] = `
"<div>
    <div data-testid="OOP1" id="OOP1" class="ad-placeholder" data-position="OOP1"></div>
</div>"
`;

exports[`<Ad /> should render the ad with correct attr when set id and position 1`] = `
"<div>
    <div data-testid="lead2" id="lead2" class="ad-placeholder" data-position="lead2"></div>
</div>"
`;
